### DISCUSSION TITLE
System - SEF - incorrect generation of canonical link

### DISCUSSION THREAD

**Discussion by Sulpher** [joomla/joomla-cms#31806] [2. Cold Features]
I am experiencing with System - SEF plugin:
<img width="738" alt="Снимок экрана 2020-12-30 в 10 38 15" src="https://user-images.githubusercontent.com/3432048/103337108-43014b80-4a8b-11eb-88ed-81beebc878e0.png">

I go to System - SEF, type my domain.
Well, all com_content articles get canonical links.
From the first look, it is logical.  But the problem is that the plugin does not generate the correct link.
E.g. we have the following page:
**articles/536-hello-world**
Let's add UTM tag:
**articles/536-hello-world?utm=test**

In fact, the plugin will generate:
```	<link href="[DOMAIN]/articles/536-hello-world?utm=test" rel="canonical" />```

Thereby, the main goal of canonical link does not working with SEF Joomla plugin.

If we open the page as:
**articles/536-hello-world?utm=test**
The canonical tag should be as follow:
**articles/536-hello-world**

Tested with Joomla 3.9.23

**Comment by infograf768** [comment]
The sef `Site Domain` parameter only purpose is to redirect to another domain with same structure/content, if desired.

<img width="731" alt="Screen Shot 2020-12-30 at 10 31 07" src="https://user-images.githubusercontent.com/869724/103342608-3af4dc00-4a8a-11eb-84aa-b0b0e8d87c37.png">

Its aim is not to avoid duplicate content urls for robots as the original "canonical" wording means.

**Comment by Sulpher** [comment]
> The sef `Site Domain` parameter only purpose is to redirect to another domain with same structure/content, if desired.
 
If this parameter has value, then canonical tag appears in the HTML source.
Then let it consider as a feature request since canonical link is a very useful option that impacts site SEO.

**Comment by simbus82** [comment]
> I am experiencing with System - SEF plugin:

Unfortunately this is one of the dangerous Joomla plugins ...
Refers to a Canonical of DOMAIN, not the usual Canonical SEO for pages.
(Why??? Who needs a domain canonicalization when for a domain migration is needed a completely different method, like a good redirect strategy?)

And yes, joomla lacks a properly functioning system for canonicals.
https://extensions.joomla.org/tags/url-canonicalization/
Here you will find some useful plugins, automatic or manual.

**Comment by infograf768** [comment]
I do agree it would be nice to have a native real canonical functionality in J.

**Comment by brianteeman** [comment]
>(Why??? Who needs a domain canonicalization when for a domain migration is needed a completely different method, like a good redirect strategy?)

It exists for people who have more than one domain name pointing to their site

**Comment by Sulpher** [comment]
> > I am experiencing with System - SEF plugin:
> 
> Unfortunately this is one of the dangerous Joomla plugins ...
> Refers to a Canonical of DOMAIN, not the usual Canonical SEO for pages.
> (Why??? Who needs a domain canonicalization when for a domain migration is needed a completely different method, like a good redirect strategy?)

I do agree: this is not what Joomla should do. There are .htaccess rules allowing to solve such a task.


> And yes, joomla lacks a properly functioning system for canonicals.
> https://extensions.joomla.org/tags/url-canonicalization/
> Here you will find some useful plugins, automatic or manual.

Thanks, but the main idea is to add such a feature as a Joomla core.

**Comment by ReLater** [comment]
> And yes, joomla lacks a properly functioning system for canonicals.
https://extensions.joomla.org/tags/url-canonicalization/
Here you will find some useful plugins, automatic or manual.

1) Why should Joomla have it in core then?

2) Did you really test them thoroughly? In years I haven't found a canonical extension that does its work well.  I only have seen more horrible and time-consuming chaos on the sites concerning SEO than benefits in the end.

I still prefer redirects whenever a "DC" link occurs in search engines. Plus a correctly configured and unambiguous menu structure. 

EDIT: We had this discussion already several times in the past.

**Comment by Sulpher** [comment]
> 1. Why should Joomla have it in core then?

In 99% of cases, canonical domain is a useless feature, at least, it is a very specific task and in most cases, there are .htaccess rules and hosting panel which does help with moving the domain. However, this plugin exists. 
Why not extend it with a more useful option? 

> 2. Did you really test them thoroughly? In years I haven't found a canonical extension that does its work well.  I only have seen more horrible and time-consuming chaos on the sites concerning SEO than benefits in the end.
> I still prefer redirects whenever a "DC" link occurs in search engines. Plus a correctly configured and unambiguous menu structure.

The site structure always should be correct, but it is not enough. 
However, as I mentioned, there might be situations when Search engine can index pages with prefixes. E.g.  UTM tag to be used for statistics or promotion campaigns.
 
https://moz.com/learn/seo/canonicalization#:~:text=A%20canonical%20tag%20(aka%20%22rel,content%20appearing%20on%20multiple%20URLs.

**Comment by ReLater** [comment]
From my point of view canonicals are unnecessary for 99.9% of pages. Theoretically Joomla knows diverse links that lead to the same page. The only way to protect your search index from "nonsense" links is to not let them occur on your site. No plugin will be able to set the correct and wanted canonical for all of these variants => chaos. It begins with the fact that any URL must be checked by Joomla to decide which canonical to set.

> E.g. UTM tag to be used for statistics or promotion campaigns.

Empirical value: I myself think there should be a solution for each individual site that uses things like that instead of establishing unreliable, performance eating solutions in core for edge cases.

The main point. You said:

> Here you will find some useful plugins, automatic or manual.

Why a solution in core then? Users that have no idea what wrong canonicals will damage will activate it blindly because a stupid SEO tool tells them "Ooops! No canonical found."

**Comment by Sulpher** [comment]
> Empirical value: I myself think there should be a solution for each individual site that uses things like that instead of establishing unreliable, performance eating solutions in core for edge cases.

It depends on how it will be implemented. 

> The main point. You said:
> 
> > Here you will find some useful plugins, automatic or manual.

It is not my quote.

> Why a solution in core then? Users that have no idea what wrong canonicals will damage will activate it blindly because a stupid SEO tool tells them "Ooops! No canonical found."

This approach is wrong. If someone has no enough skills, it is not a reason to stop improving CMS. 
There might be a tooltip description and a URL to read what canonical links are.

As for indexing, there is not only Google.
Here is Yandex search engine (Russian) which indicates that the page has no canonical link.
<img width="891" alt="Снимок экрана 2020-12-30 в 17 05 59" src="https://user-images.githubusercontent.com/3432048/103356219-6941de00-4ac1-11eb-96a9-0bdc94791227.png">

**Comment by ReLater** [comment]
Simple: Start programming or at least defining! A canonical plugin/extension in core must match all needs of all users/use cases.

If Yandex or SEO tools suggest errors with their "non-canonical" marks that shouldn't be a measure for Joomla as long it can't provide a reliable and perfect solution.

Just setting canonicals is not a protection against DC. You simply don't see these warnings any more.

**Comment by ReLater** [comment]
But we are away from your issue that seems to be a warrantable.

**Comment by simbus82** [comment]
> From my point of view canonicals are unnecessary for 99.9% of pages. 

It's your point of view, ok. But it is wrong.

> Why a solution in core then? Users that have no idea what wrong canonicals will damage will activate it blindly because a stupid SEO tool tells them "Ooops! No canonical found."

No no... don't hide behind the "unskilled user" problem.

Canonical are needed in 2020, for:
- different "surfaces" where website can show content: amp, mobile, display, etc.
- cleaning of dinamical urls (shop & filters for example)
- many url (SEF, not-SEF, multi-category, aliasing) that referer to same page (JOOMLA!!!!!!!!)
- protection from bad backlinking (www, not-www, http/https, etc. remember how Search Engine bot/spiders works)
- syndication of posts and content
- crawl budget with big webites
- cannibalization protection for webpages that cannot be redirected
- etc, etc.

> Theoretically Joomla knows diverse links that lead to the same page. The only way to protect your search index from "nonsense" links is to not let them occur on your site. No plugin will be able to set the correct and wanted canonical for all of these variants => chaos. It begins with the fact that any URL must be checked by Joomla to decide which canonical to set.

Then, is it difficult to manage Joomla Canonicals? **Yes, it is!**
But a simple field in the article and menu manager can be really helpful for example, so we can setup it manually.
In the future someone will can think about developping an automatic feature.

**Comment by Sulpher** [comment]
@ReLater
> Simple: Start programming or at least defining! A canonical plugin/extension in the core must match all needs of all users/use cases.

Listen, have you learned such a definition as a polite discussion? If you took a responsibility to "teach" me what to do, I rather would say you should google what are canonical URLs first.
https://yoast.com/rel-canonical/

@simbus82 has already given the best example above why it's important.

> If Yandex or SEO tools suggest errors with their "non-canonical" marks that shouldn't be a measure for Joomla as long it can't provide a reliable and perfect solution.

Russian is spoken by about 300 millions people and there are so many websites made with Joomla and since OSM does the best to cover all possible communities, then counting such a request could positively impact Joomla reputation. 
But it's even not about Yandex, the question is about improving the existing plugin.
I just shared my thoughts since Joomla is the community and users' feedbacks matters, so as we can see, I am not the only person who thinks canonical links are a must-have.

**Comment by alikon** [comment]
can someone  move this in https://github.com/joomla/joomla-cms/discussions/categories/ideas 
it's ok ?

**Comment by infograf768** [comment]
Done.
